File: DifferentiablePointRender/GaussianSplatting/src/compute_sh.cpp

#include <utils.h>
#include <compute_sh.h>
#include <compute_sh_kernel.h>

std::tuple<torch::Tensor, torch::Tensor> 
computeSHForward(
    const torch::Tensor& shs,
    const int degree,
    const torch::Tensor& view_dirs,
    const torch::Tensor& visibility_status
){
    CHECK_INPUT(shs);
    CHECK_INPUT(view_dirs);
    CHECK_INPUT(visibility_status);

    const int P = shs.size(0);
    auto float_opts = shs.options().dtype(torch::kFloat32);
    auto bool_ops = shs.options().dtype(torch::kBool);
    torch::Tensor colors = torch::zeros({P, 3}, float_opts);
    torch::Tensor clamped = torch::ones({P, 3}, bool_ops);

    if(P != 0)
    {
        computeSHForwardCUDA(
            P, 
            shs.contiguous().data_ptr<float>(), 
            degree, 
            view_dirs.contiguous().data_ptr<float>(), 
            visibility_status.contiguous().data_ptr<bool>(), 
            colors.data_ptr<float>(), 
            clamped.data_ptr<bool>()
        );
    }
    
    return std::make_tuple(colors, clamped);
}

std::tuple<torch::Tensor, torch::Tensor> 
computeSHBackward(
    const torch::Tensor& shs,
    const int degree,
    const torch::Tensor& view_dirs,
    const torch::Tensor& visibility_status,
    const torch::Tensor& clamped,
    const torch::Tensor& dL_dcolors
){
    CHECK_INPUT(shs);
    CHECK_INPUT(view_dirs);
    CHECK_INPUT(visibility_status);
    CHECK_INPUT(dL_dcolors);

    const int P = shs.size(0);
    const int S = shs.size(1);
    auto float_opts = shs.options().dtype(torch::kFloat32);
    torch::Tensor dL_dshs = torch::zeros({P, S, 3}, float_opts);
    torch::Tensor dL_dvdirs = torch::zeros({P, 3}, float_opts);

    if(P != 0)
    {
        computeSHBackwardCUDA(
            P, 
            shs.contiguous().data_ptr<float>(), 
            degree, 
            view_dirs.contiguous().data_ptr<float>(), 
            visibility_status.contiguous().data_ptr<bool>(),
            clamped.contiguous().data_ptr<bool>(),
            dL_dcolors.contiguous().data_ptr<float>(), 
            dL_dshs.data_ptr<float>(),
            dL_dvdirs.data_ptr<float>()
        );
    }
    
    return std::make_tuple(dL_dshs, dL_dvdirs);
}

File: DifferentiablePointRender/GaussianSplatting/compute_sh.py

import torch
from torch import Tensor
from typing import Tuple
from jaxtyping import Float, Bool
import DifferentiablePointRender.GaussianSplatting._C as _C


def compute_sh(
    shs,
    degree,
    view_dirs,
    visibility_status=None
)->Tensor:
    
    if visibility_status is None:
        visibility_status = torch.ones_like(shs[:, 0, 0], dtype=torch.bool)
    
    return _ComputeSH.apply(
        shs,
        degree,
        view_dirs,
        visibility_status
    )
    

class _ComputeSH(torch.autograd.Function):
    @staticmethod
    def forward(
        ctx,
        shs,
        degree,
        view_dirs,
        visibility_status
    ):
        
        (
            color,
            clamped
        ) = _C.compute_sh_forward(
            shs,
            degree,
            view_dirs,
            visibility_status
        )
        
        # save variables for backward
        ctx.degree = degree
        ctx.save_for_backward(
            shs, 
            view_dirs, 
            visibility_status,
            clamped
        )
        
        return color

    @staticmethod
    def backward(
        ctx, 
        dL_dcolor):
        
        # get saved variables from forward
        degree = ctx.degree
        (
            shs, 
            view_dirs, 
            visibility_status, 
            clamped
        ) = ctx.saved_tensors
        
        (
            dL_dshs, 
            dL_dvdirs
        ) = _C.compute_sh_backward(
            shs,
            degree,
            view_dirs,
            visibility_status,
            clamped,
            dL_dcolor
        )
        
        grads = (
            # loss gradient w.r.t shs
            dL_dshs,
            # loss gradient w.r.t dgree
            None,
            # loss gradient w.r.t view_dirs
            dL_dvdirs,
            # loss gradient w.r.t visibility_status
            None,
        )

        return grads
